runtime/interpreter/interpreter.go

```diff
@@ -1905,10 +1988,88 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
 			return ConvertAddress(interpreter, value, locationRange)
 		}
 
+	case *sema.ConstantSizedType, *sema.VariableSizedType:
+		if arrayValue, isArray := value.(*ArrayValue); isArray && !valueType.Equal(unwrappedTargetType) {
+
+			arrayStaticType := interpreter.convertStaticType(arrayValue.StaticType(interpreter), unwrappedTargetType).(ArrayStaticType)
+			targetElementType := interpreter.MustConvertStaticToSemaType(arrayStaticType.ElementType())
+
+			values := make([]Value, 0, arrayValue.Count())
+
+			arrayValue.Iterate(interpreter, func(v Value) bool {
+				valueType := interpreter.MustConvertStaticToSemaType(v.StaticType(interpreter))
+				values = append(values, interpreter.convert(v, valueType, targetElementType, locationRange))
+				return true
+			})
+
+			return NewArrayValue(
+				interpreter,
+				locationRange,
+				arrayStaticType,
+				arrayValue.GetOwner(),
+				values...,
```

Comment:
Instead of allocating a temporary slice, `NewArrayValueWithIterator` can be used to construct the new array value  from an existing "iterator". Same below for dictionary values.